components/brave_rewards/resources/extension/brave_rewards/background/publisherVisit.ts

```diff
@@ -0,0 +1,250 @@
+      return
+    })
+}
+
+const savePublisherInfo = (tabId: number, mediaType: string, url: string, publisherKey: string, publisherName: string, favIconUrl: string) => {
+  chrome.braveRewards.savePublisherInfo(
+    tabId,
+    mediaType,
+    url,
+    publisherKey,
+    publisherName,
+    favIconUrl,
+    (result: RewardsExtension.Result) => {
+      if (result !== 0) {
+        console.error(`Failed to save publisher info for ${publisherKey}, result is ${result}`)
+        return
+      }
+    })
+}
+
+const handleSavePublisherVisitResponse = (tabId: number, mediaType: string, data: SavePublisherVisitResponse) => {
+  if (!data.publisherKey) {
+    console.error('Failed to handle publisher visit: missing publisher key')
+    return
+  }
+
+  if (data.mediaKey && !publisherKeys[data.mediaKey]) {
+    publisherKeys[data.mediaKey] = data.publisherKey
+  }
+
+  chrome.braveRewards.getPublisherInfo(
+    data.publisherKey, (result: RewardsExtension.Result, info?: RewardsExtension.Publisher) => {
+      if (result === 0 && info) {
+        getPublisherPanelInfo(tabId, data.publisherKey)
+        return
+      }
+
+      // Failed to find publisher info corresponding to this key, so save it now
+      if (result === 9) {
+        savePublisherInfo(
+          tabId,
+          mediaType,
+          data.url,
+          data.publisherKey,
+          data.publisherName,
+          data.favIconUrl || '')
+        return
+      }
+    })
+}
+
+const processGreaselionMessage = (msg: any, sender: chrome.runtime.MessageSender) => {
+  if (!msg || !sender || !sender.tab) {
+    return
+  }
+
+  const windowId = sender.tab.windowId
+  if (!windowId) {
+    return
+  }
+
+  const tabId = sender.tab.id
+  if (!tabId) {
+    return
+  }
+
+  const response = msg as GreaselionResponse
+  if (!response.data) {
+    console.error(`Received empty Greaselion response payload for ${msg.type} message`)
+    return
+  }
+
+  switch (msg.type) {
+    case 'GreaselionError': {
+      const data = response.data as GreaselionErrorResponse
+      handleGreaselionErrorResponse(tabId, response.mediaType, data)
+      break
+    }
+    case 'SavePublisherVisit': {
+      const data = response.data as SavePublisherVisitResponse
+      handleSavePublisherVisitResponse(tabId, response.mediaType, data)
+      break
+    }
+  }
+}
+
+chrome.runtime.onConnectExternal.addListener((port: chrome.runtime.Port) => {
+  if (!port || !port.sender || !port.sender.id || !port.sender.tab || port.name !== 'Greaselion') {
+    return
+  }
+
+  const tabId = port.sender.tab.id
+  if (!tabId) {
+    return
+  }
+
+  if (!greaselionPorts.get(tabId)) {
+    greaselionPorts.set(tabId, port)
+  }
+
+  port.onMessage.addListener((msg: any, port: chrome.runtime.Port) => {
+    if (!port.sender || !port.sender.id || !msg) {
+      return
+    }
+
+    chrome.greaselion.isGreaselionExtension(port.sender.id, (valid: boolean) => {
+      if (!valid) {
+        return
+      }
+      switch (msg.type) {
```

Comment:
I think that we already talked about this one, but why do we need to send this to GL?

Reply:
That's receiving, not sending. Greaselion sends a registration request to the background page with the media duration url pattern. The background page listens for the completed web requests because content scripts don't have access to that API. When the request completes, we send it to Greaselion to parse the duration from it.